src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Download/PackageDownloadRunner.cs

```diff
@@ -188,6 +216,45 @@ await ResolvePackageDownloadVersion(
             return (versionToDownload, downloadSourceRepository);
         }
 
+        /// <summary>
+        /// Builds the set of SourceRepository objects to use for a given package,
+        /// applying package source mapping
+        /// validating HTTP usage only on the *effective* sources.
+        /// </summary>
+        private static bool TryGetRepositoriesForPackage(
+            string packageId,
+            PackageDownloadArgs args,
+            PackageSourceMapping packageSourceMapping,
+            IReadOnlyDictionary<string, SourceRepository> sourceRepositoriesMap,
+            List<SourceRepository> allRepos,
+            ILoggerWithColor logger,
+            out List<SourceRepository> repositories)
+        {
+            var mappedNames = packageSourceMapping.GetConfiguredPackageSources(packageId);
+            var mappedRepos = mappedNames
+                .Select(name => sourceRepositoriesMap[name])
```

Comment:
This can fail. 
There's no coherence guarantee between available sources and mapped sources, so you need a different way of doing this. 

See: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteWalkContext.cs#L78-L93